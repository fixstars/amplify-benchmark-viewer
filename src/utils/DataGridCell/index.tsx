import { LinkButton } from 'components/atoms/LinkButton'

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    let backgroundColor = '#FF66FF'

    switch (client) {
      case 'FixstarsClient':
        backgroundColor = '#0F66DC'
        break
      case 'ToshibaClient':
        backgroundColor = '#f57bae'
        break
      case 'ToshibaSQBM2Client':
        backgroundColor = '#e75e75'
        break
      case 'GurobiClient':
        backgroundColor = '#e5f97c'
        break
      case 'DWaveSamplerClient':
        backgroundColor = '#1b000d'
        break
      case 'LeapHybridSamplerClient':
        backgroundColor = '#f6a695'
        break
      case 'FujitsuDASolverClient':
        backgroundColor = '#692981'
        break
      case 'FujitsuDASolverExpertClient':
        backgroundColor = '#3cde29'
        break
      case 'FujitsuDAPTSolverClient':
        backgroundColor = '#591414'
        break
      case 'FujitsuDAMixedModeSolverClient':
        backgroundColor = '#d66594'
        break
      case 'FujitsuDA2SolverClient':
        backgroundColor = '#2d884a'
        break
      case 'FujitsuDA2SolverExpertClient':
        backgroundColor = '#400763'
        break
      case 'FujitsuDA2PTSolverClient':
        backgroundColor = '#9d19f4'
        break
      case 'FujitsuDA2MixedModeSolverClient':
        backgroundColor = '#fcdeb9'
        break
      case 'FujitsuDA3SolverClient':
        backgroundColor = '#718de6'
        break
      case 'FujitsuDA4SolverClient':
        backgroundColor = '#1f4c48'
        break
      case 'HitachiClient':
        backgroundColor = '#909743'
        break
      case 'ABSClient':
        backgroundColor = '#2dc1c8'
        break
      case 'QiskitClient':
        backgroundColor = '#c17913'
        break
      case 'QulacsClient':
        backgroundColor = '#b2a1f6'
        break
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (client: string): string => {
    let backgroundColor = '#FF66FF'

    switch (client) {
      case 'pr124':
        backgroundColor = '#0F66DC'
        break
      case 'pr136':
        backgroundColor = '#f57bae'
        break
      case 'pr144':
        backgroundColor = '#e75e75'
        break
      case 'si175':
        backgroundColor = '#e5f97c'
        break
      case 'pr299':
        backgroundColor = '#1b000d'
        break
      case 'sko56':
        backgroundColor = '#f6a695'
        break
      case 'sko100a':
        backgroundColor = '#692981'
        break
      case 'tai150b':
        backgroundColor = '#3cde29'
        break
      case 'G65':
        backgroundColor = '#591414'
        break
      case 'G77':
        backgroundColor = '#d66594'
        break
      case 'G81':
        backgroundColor = '#2d884a'
        break
      case '16x16_1_000':
        backgroundColor = '#400763'
        break
      case '25x25_000':
        backgroundColor = '#9d19f4'
        break
      case 'B-n31-k5':
        backgroundColor = '#fcdeb9'
        break
      case 'F-n45-k4':
        backgroundColor = '#718de6'
        break
      case 'burma14':
        backgroundColor = '#1f4c48'
        break
      case 'ulysses16':
        backgroundColor = '#909743'
        break
      case 'bayg29':
        backgroundColor = '#2dc1c8'
        break
      case 'kroA100':
        backgroundColor = '#c17913'
        break
      case 'esc32a':
        backgroundColor = '#b2a1f6'
        break
      case 'G1':
        backgroundColor = '#9a916d'
        break
      case 'G11':
        backgroundColor = '#406a0f'
        break
      case 'G32':
        backgroundColor = '#004b31'
        break
      case 'K2000':
        backgroundColor = '#18729f'
        break
      case '9x9_h39_022':
        backgroundColor = '#9997c3'
        break
      case '9x9_h17_055':
        backgroundColor = '#9cf0ec'
        break
      case '16x16_easy_233':
        backgroundColor = '#ce843c'
        break
      case '16x16_hard_194':
        backgroundColor = '#2d326e'
        break
      case 'P-n16-k8':
        backgroundColor = '#76ed60'
        break
      case 'P-n19-k2':
        backgroundColor = '#d7f7e8'
        break
      case 'tai256c':
        backgroundColor = '#bb15d7'
        break
      case 'eil51':
        backgroundColor = '#87a7d8'
        break
      case 'A-n32-k5':
        backgroundColor = '#94053a'
        break
      case 'chr12a':
        backgroundColor = '#e5c8a6'
        break
      case 'chr18a':
        backgroundColor = '#e2f2c0'
        break
    }

    return backgroundColor
  }

  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      {problemInstances[className]?.map((instance) => (
        <div
          key={`${key}_${className}_${instance}`}
          style={{ marginRight: 10 }}
        >
          <LinkButton
            label={instance}
            link={`/classes/${className}/instances/${instance}`}
            backgroundColor={getBackgroundColor(instance)}
          />
        </div>
      ))}
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
