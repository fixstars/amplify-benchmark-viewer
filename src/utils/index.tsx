export * from './DataGridCell'
export * from './statistics'
