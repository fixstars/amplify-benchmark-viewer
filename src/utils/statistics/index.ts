function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
