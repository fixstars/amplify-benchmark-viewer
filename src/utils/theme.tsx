import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
