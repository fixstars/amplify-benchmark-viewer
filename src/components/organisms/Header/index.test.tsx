import { fireEvent, render, screen } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { BrowserRouter, Router } from 'react-router-dom'
import { mockAppBar } from 'utils/test'

import { Header } from '.'

describe('<Header />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>,
    )

    const appBar = mockAppBar.mock.calls[0][0]
    expect(appBar.position).toBe('fixed')
    expect(appBar.elevation).toBe(0)

    const toolbar = appBar.children
    expect(toolbar.type.name).toBe('Toolbar')
    expect(toolbar.props.sx).toEqual({ flexWrap: 'wrap' })

    const headerContents = toolbar.props.children
    expect(headerContents.length).toBe(2)

    const appTitleLinkContainer = headerContents[0]
    expect(appTitleLinkContainer.type.render.displayName).toBe('Styled(div)')
    const appTitleLink = appTitleLinkContainer.props.children
    expect(appTitleLink.type.render.displayName).toBe('Styled(Link)')
    expect(appTitleLink.props.to).toBe('.')

    const appTitle = appTitleLink.props.children
    expect(appTitle.type.name).toBe('Typography')
    expect(appTitle.props.variant).toBe('h6')
    expect(appTitle.props.color).toBe('inherit')
    expect(appTitle.props.noWrap).toBe(true)
    expect(appTitle.props.sx).toEqual({ color: '#fff' })
    expect(appTitle.props.children).toBe('Annealing Benchmark')

    const navLinks = headerContents[1]
    expect(navLinks.type.render.displayName).toBe('Styled(nav)')
    const links = navLinks.props.children
    expect(links.length).toBe(3)

    {
      const link = links[0]
      expect(link.type.name).toBe('NavLink')
      expect(link.props).toEqual({
        url: '/',
        label: 'Problems',
        isActive: true,
      })
    }

    {
      const link = links[1]
      expect(link.type.name).toBe('NavLink')
      expect(link.props).toEqual({
        url: '/clients',
        label: 'Clients',
        isActive: false,
      })
    }

    {
      const link = links[2]
      expect(link.type.name).toBe('NavLink')
      expect(link.props).toEqual({
        url: '/labels',
        label: 'Label',
        isActive: false,
      })
    }

    expect(container).toMatchSnapshot()
  })

  it('Go to the root page when the app title is clicked', () => {
    const mockPush = jest.fn()
    const history = createMemoryHistory({ initialEntries: ['/pages/1'] })
    history.push = mockPush

    render(
      <Router location={history.location} navigator={history}>
        <Header />
      </Router>,
    )

    fireEvent.click(screen.getByText('Annealing Benchmark'))
    expect(mockPush.mock.calls[0][0].pathname).toBe('/')
  })

  it('Go to the specific page when nav menu is clicked', () => {
    const mockPush = jest.fn()
    const history = createMemoryHistory({ initialEntries: ['/'] })
    history.push = mockPush

    render(
      <Router location={history.location} navigator={history}>
        <Header />
      </Router>,
    )

    jest.clearAllMocks()

    fireEvent.click(screen.getByText('Clients'))
    expect(mockPush.mock.calls[0][0].pathname).toBe('/clients')

    jest.clearAllMocks()

    fireEvent.click(screen.getByText('Label'))
    expect(mockPush.mock.calls[0][0].pathname).toBe('/labels')
  })
})
