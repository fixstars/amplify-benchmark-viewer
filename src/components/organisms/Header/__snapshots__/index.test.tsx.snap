// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Rendered well 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1hc7nu0-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
    >
      <div
        class="css-yp9swi"
      >
        <a
          class="css-94l1l1"
          href="/"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
          >
            Annealing Benchmark
          </h6>
        </a>
      </div>
      <nav
        class="css-ov1ktg"
      >
        <a
          class="css-94l1l1"
          href="/"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
          >
            Problems
          </p>
        </a>
        <a
          class="css-94l1l1"
          href="/clients"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
          >
            Clients
          </p>
        </a>
        <a
          class="css-94l1l1"
          href="/labels"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
          >
            Label
          </p>
        </a>
      </nav>
    </div>
  </header>
</div>
`;
